### ISSUE TITLE
Implement download images on demand.

### ISSUE THREAD

**Issue by lodici** [magarena/magarena#713]
@neoedmund has as pull request (see #709) that will download images on demand. It is a surprisingly small amount of code, in affect it updates the card selection rule to 
```
custom -> crops -> cards/tokens -> download -> missing
```

However, need to clarify how it affects the current implementation. For example, the new images alert would seem to be redundant. I am fairly sure it will not work for non-english cards or when images are flagged as updated, without some tweaking anyway. Will it have an affect on performance if the internet connection is slow or down?

To cater for CardBuilder, most image loading is now done on background threads which means the download on demand has no obvious adverse effects on explorer or the deck editor. When starting a new duel I believe it will automatically download the images needed during the startup background thread, but I need to check.

I had a play on my lowly acer and the fact that you can start playing immediately is a big usability improvement so maybe it is worth considering adding to 1.71? @melvinzhang, @ShawnieBoy your thoughts?

**Comment by ShawnieBoy** [comment]
I think it's a great idea - if the internet is down it would go to -> missing and then proxies made. Maybe a check for download time and kick in proxies if it's taking a long time.

Why wouldn't language and image_updated be caught?

**Comment by lodici** [comment]
> Why wouldn't language and image_updated be caught?

As submitted it is not checking for these but I think we can tweak it to do so.

> Maybe a check for download time and kick in proxies if it's taking a long time.

Actually I tested it with no connection and it triggered the missing proxies as expected. However connection timeout and knowing when to stop trying becomes more important when the downloading is incidental to what the user is trying to do. When running from the download dialog you can always cancel if it is generating loads of errors or taking too long.

**Comment by lodici** [comment]
_Card language_ : is stored as general.config so should be ok.
_image_updated_ : is more tricky. the solution I think is that at startup identify all cards that are out-of-date and delete the existing image. Thus it will then be flagged as missing and so can be downloaded either on demand or via the download dialog.
_no connection_ : fails immediately. Which is a good thing.
_slow connection_ :  timeout is currently set to 10 seconds for connection and 5 seconds for read. Somehow, need to test how it would behave where downloads are crawling for some reason with regular timeouts. Worst case is that it takes 5 seconds for each image before giving up and creating a missing proxy. That could delay the start of a duel, tiled cards screen beyond acceptable duration.

**Comment by melvinzhang** [comment]
One way to prevent slow downloads from interrupting the game is to decouple the downloading from the display, i.e. first show the proxy, then add the download to a queue, once the download is complete replace the image in the cache with the downloaded image.

This change also assume the user always wants to see the downloaded card image, what if the user only wants to use proxy? For example, because there is not enough space to store all the images. Right now I can do that by never clicking on image download.

**Comment by melvinzhang** [comment]
This download queue should be managed by DownloadWorker so we can use the same logic as for downloading  via the DownloadPanel.

**Comment by lodici** [comment]
> This change also assume the user always wants to see the downloaded card image, what if the user only wants to use proxy? For example, because there is not enough space to store all the images. Right now I can do that by never clicking on image download.

It is easy enough to add a setting to make it optional although I feel that it should be on by default when installing Magarena for the first time. If importing an existing version and the setting does not yet exist then it should be off by default to match current behaviour. Also, the new images alert will only be shown if the setting is off.

> One way to prevent slow downloads from interrupting the game is to decouple the downloading from the display, i.e. first show the proxy, then add the download to a queue, once the download is complete replace the image in the cache with the downloaded image.

I don't think it needs to be as complicated as this. I did start looking into using a queue a while back but since then have implemented CardImageWorker for getting images on a background thread, ostensibly to cater for CardBuilder startup, but which also works with the download on demand code since it runs in the same worker thread. This is the worker that lets you zip through cards in the deck editor or explorer without being affected by image loading.

I have broadband and have tested @neoedmund's basic update and for most screens unless I knew I would not even realize the images are being downloaded before being displayed. That is pretty cool! The only screen where there is a relatively longer wait than normal is the deck tiled screen and even then it has to be a large deck. When running a new duel using random decks, there is no appreciable delay - most noticeable is a slight pause if you look at the AI's hand.

From this discussion it is clear that it is not ready for 1.71 but perhaps can be introduced (as devMode?) in 1.72 so we have a month to tweak and stress test it.

**Comment by neoedmund** [comment]
For image update, I have a solution tested but not pushed yet. It passes any callee jpanel as parameter, when downloaded, in other thread, clear the image cache with the key of the card and repaint the panel. It workes great.

**Comment by lodici** [comment]
I have added a ``download images on demand`` setting which basically implements @neoedmund's light and elegant original suggestion (not the massive set of changes subsequently pushed). It is switched off by default on account of it being a bit experimental but I think even as it stands most people would find this a more convenient option, unless they want to use proxy frames with cropped images or non-English cards. So I think we should be considering it as a default option or at least we should probably ask the user's preference on running a new install. If switched on then the new images alert is suppressed and access to the download images dialog is prevented. What I also like about this option is that it means we will not be hammering ``magicards.info`` which may or may not be the cause of #729 as raised by @slartifer. It can't hurt, anyway.

**Comment by ShawnieBoy** [comment]
Wouldn't the download on demand work with non-English and crops anyway (and unimplemented)?

I'll just go have a play :grinning:

**Comment by lodici** [comment]
> Wouldn't the download on demand work with non-English and crops anyway (and unimplemented)?

I think it should but initially I just wanted to get the functionality in there that required the least amount of effort on my part!

Currently, the decision to use crops or foreign language cards is only made in the downloads dialog prior to running the download. We would need to move this choice somewhere else and save that choice persistently, probably as additional settings in preferences but also I think we would need to add a screen when running Magarena the first time after an install - eg, do you want to use images on demand? and if YES show (a) original images, (b) proxy frames with cropped images, (c) non-English images and if (c) select your preferred language, etc.

**Comment by ShawnieBoy** [comment]
I'm with you!

Runs perfectly btw - even doing lots of searches in Card Explorer.

**Comment by lodici** [comment]
Great, I have the same experience (which was why I was keen to get something published) - if I didn't know I would not even realize that the images were being downloaded first. Still needs refining (eg. ``image_updated`` is not currently supported) but at least it is another option especially if ``magicards.info`` decides to play hardball.

**Comment by ShawnieBoy** [comment]
It might be worth trying to get in contact with the curator of magiccard.info in case we're causing them issues or if there is a preferred way for their servers to handle requests.

**Comment by lodici** [comment]
Not sure if I want to put my head above the parapet! :boom:

**Comment by lodici** [comment]
Option added, create new issue for further updates.
